File: src/services/compliance/notices.ts
import { prisma } from "@/lib/db";
import { generateResponse } from "@/lib/ai";
import { sendEmail } from "@/lib/email";
import {
  NOTICE_LETTER_SYSTEM_PROMPT,
  buildNoticeLetterPrompt,
  type NoticeLetterContext,
} from "./prompts";
import type {
  ComplianceNotice,
  ComplianceNoticeType,
  ComplianceNoticeStatus,
  ContractClauseMethod,
} from "@prisma/client";

// ── Types ─────────────────────────────────────────────────────────────────

interface CreateNoticeParams {
  projectId: string;
  deadlineId: string;
  clauseId: string;
  recipientName: string;
  recipientEmail?: string;
  generateWithAI?: boolean;
  createdById: string;
}

interface DeliveryConfirmation {
  email?: {
    sentAt: string;
    status: string;
    deliveredAt?: string;
  };
  certifiedMail?: {
    sentAt: string;
    trackingNumber: string;
    carrier: string;
    status: string;
    deliveredAt?: string;
    signedBy?: string;
  };
  handDelivery?: {
    deliveredAt: string;
    receivedBy: string;
  };
  courier?: {
    sentAt: string;
    trackingNumber: string;
    carrier: string;
    status: string;
    deliveredAt?: string;
  };
}

// ── Map clause kind → notice type ─────────────────────────────────────────

const CLAUSE_TO_NOTICE_TYPE: Record<string, ComplianceNoticeType> = {
  CLAIMS_PROCEDURE: "CLAIM_NOTICE",
  CHANGE_ORDER_PROCESS: "CHANGE_ORDER_NOTICE",
  TERMINATION: "TERMINATION_NOTICE",
  WARRANTY: "WARRANTY_NOTICE",
  NOTICE_REQUIREMENTS: "CLAIM_NOTICE",
  DISPUTE_RESOLUTION: "CLAIM_NOTICE",
  PAYMENT_TERMS: "NOTICE_TO_PROCEED",
  RETENTION: "LIEN_NOTICE",
};

function clauseKindToNoticeType(kind: string): ComplianceNoticeType {
  return CLAUSE_TO_NOTICE_TYPE[kind] || "CLAIM_NOTICE";
}

// ── Map clause noticeMethod → delivery methods ────────────────────────────

function noticeMethodToDeliveryMethods(
  method: ContractClauseMethod | null
): string[] {
  switch (method) {
    case "CERTIFIED_MAIL":
      return ["CERTIFIED_MAIL", "EMAIL"];
    case "REGISTERED_MAIL":
      return ["REGISTERED_MAIL", "EMAIL"];
    case "HAND_DELIVERY":
      return ["HAND_DELIVERY", "EMAIL"];
    case "EMAIL":
      return ["EMAIL"];
    case "WRITTEN_NOTICE":
      return ["EMAIL", "CERTIFIED_MAIL"];
    default:
      return ["EMAIL", "CERTIFIED_MAIL"];
  }
}

// ── Create Notice ─────────────────────────────────────────────────────────

export async function createNotice(
  params: CreateNoticeParams
): Promise<ComplianceNotice> {
  const {
    projectId,
    deadlineId,
    clauseId,
    recipientName,
    recipientEmail,
    generateWithAI = true,
    createdById,
  } = params;

  // Fetch clause + deadline + project + org (validate project ownership)
  const clause = await prisma.contractClause.findFirst({
    where: { id: clauseId, projectId },
  });
  if (!clause) throw new Error("Clause not found in this project");

  const deadline = await prisma.complianceDeadline.findFirst({
    where: { id: deadlineId, projectId },
  });
  if (!deadline) throw new Error("Deadline not found in this project");

  const project = await prisma.project.findUnique({
    where: { id: projectId },
    include: { organization: { select: { name: true } } },
  });
  if (!project) throw new Error("Project not found");

  // Generate letter content with AI
  let content = "";
  let aiModel: string | null = null;

  if (generateWithAI) {
    const ctx: NoticeLetterContext = {
      sectionRef: clause.sectionRef,
      clauseKind: clause.kind,
      clauseTitle: clause.title,
      deadlineDays: clause.deadlineDays ?? 0,
      deadlineType: clause.deadlineType ?? "CALENDAR_DAYS",
      trigger: clause.trigger,
      noticeMethod: clause.noticeMethod,
      triggerEventType: deadline.triggerEventType,
      triggerDescription: deadline.triggerDescription,
      eventDate: deadline.triggeredAt.toISOString().split("T")[0],
      projectName: project.name,
      orgName: project.organization.name,
      recipientName,
      recipientEmail: recipientEmail ?? undefined,
    };

    const aiResponse = await generateResponse({
      systemPrompt: NOTICE_LETTER_SYSTEM_PROMPT,
      userPrompt: buildNoticeLetterPrompt(ctx),
      model: "sonnet",
      maxTokens: 2048,
      temperature: 0.2,
    });

    content = aiResponse.content;
    aiModel = aiResponse.model;
  }

  const noticeType = clauseKindToNoticeType(clause.kind);
  const deliveryMethods = noticeMethodToDeliveryMethods(clause.noticeMethod);

  // Create notice
  const notice = await prisma.complianceNotice.create({
    data: {
      projectId,
      type: noticeType,
      status: "DRAFT",
      title: `${clause.title} — ${clause.sectionRef || "Notice"}`,
      content,
      recipientName,
      recipientEmail: recipientEmail || null,
      dueDate: deadline.calculatedDeadline,
      clauseId,
      deliveryMethods,
      generatedByAI: generateWithAI,
      aiModel,
      createdById,
    },
  });

  // Link notice to deadline
  await prisma.complianceDeadline.update({
    where: { id: deadlineId },
    data: {
      status: "NOTICE_DRAFTED",
      noticeId: notice.id,
      noticeCreatedAt: new Date(),
    },
  });

  // Audit log
  await prisma.complianceAuditLog.create({
    data: {
      projectId,
      eventType: "NOTICE_DRAFTED",
      entityType: "ComplianceNotice",
      entityId: notice.id,
      userId: createdById,
      actorType: generateWithAI ? "AI" : "USER",
      action: "created",
      details: JSON.parse(
        JSON.stringify({
          sectionRef: clause.sectionRef,
          noticeType,
          generatedByAI: generateWithAI,
          deliveryMethods,
        })
      ),
    },
  });

  return notice;
}

// ── Regenerate Letter ─────────────────────────────────────────────────────

export async function regenerateNoticeLetter(
  noticeId: string,
  customInstructions?: string
): Promise<{ content: string }> {
  const notice = await prisma.complianceNotice.findUnique({
    where: { id: noticeId },
  });
  if (!notice) throw new Error("Notice not found");
  if (!notice.clauseId) throw new Error("Notice has no linked clause");

  const clause = await prisma.contractClause.findUnique({
    where: { id: notice.clauseId },
  });
  if (!clause) throw new Error("Clause not found");

  // Find linked deadline
  const deadline = await prisma.complianceDeadline.findFirst({
    where: { noticeId },
  });

  const project = await prisma.project.findUnique({
    where: { id: notice.projectId },
    include: { organization: { select: { name: true } } },
  });

  const ctx: NoticeLetterContext = {
    sectionRef: clause.sectionRef,
    clauseKind: clause.kind,
    clauseTitle: clause.title,
    deadlineDays: clause.deadlineDays ?? 0,
    deadlineType: clause.deadlineType ?? "CALENDAR_DAYS",
    trigger: clause.trigger,
    noticeMethod: clause.noticeMethod,
    triggerEventType: deadline?.triggerEventType ?? "OTHER",
    triggerDescription:
      deadline?.triggerDescription ?? "See attached documentation",
    eventDate:
      deadline?.triggeredAt.toISOString().split("T")[0] ??
      new Date().toISOString().split("T")[0],
    projectName: project?.name ?? "[PROJECT NAME]",
    orgName: project?.organization.name ?? "[ORGANIZATION]",
    recipientName: notice.recipientName ?? "[RECIPIENT]",
    recipientEmail: notice.recipientEmail ?? undefined,
  };

  let prompt = buildNoticeLetterPrompt(ctx);
  if (customInstructions) {
    prompt += `\n\nADDITIONAL INSTRUCTIONS: ${customInstructions}`;
  }

  const aiResponse = await generateResponse({
    systemPrompt: NOTICE_LETTER_SYSTEM_PROMPT,
    userPrompt: prompt,
    model: "sonnet",
    maxTokens: 2048,
    temperature: 0.2,
  });

  await prisma.complianceNotice.update({
    where: { id: noticeId },
    data: {
      content: aiResponse.content,
      generatedByAI: true,
      aiModel: aiResponse.model,
    },
  });

  return { content: aiResponse.content };
}

// ── Update Notice Content ─────────────────────────────────────────────────

export async function updateNoticeContent(
  noticeId: string,
  content: string,
  userId: string
): Promise<ComplianceNotice> {
  const notice = await prisma.complianceNotice.update({
    where: { id: noticeId },
    data: {
      content,
      reviewedBy: userId,
      reviewedAt: new Date(),
    },
  });

  await prisma.complianceAuditLog.create({
    data: {
      projectId: notice.projectId,
      eventType: "NOTICE_EDITED",
      entityType: "ComplianceNotice",
      entityId: noticeId,
      userId,
      actorType: "USER",
      action: "updated",
      details: JSON.parse(JSON.stringify({ field: "content" })),
    },
  });

  return notice;
}

// ── Send Notice ───────────────────────────────────────────────────────────

export async function sendNotice(
  noticeId: string,
  methods: string[],
  userId: string
): Promise<{ sentMethods: string[]; failedMethods: string[] }> {
  const notice = await prisma.complianceNotice.findUnique({
    where: { id: noticeId },
  });
  if (!notice) throw new Error("Notice not found");

  // Fetch user + org for email sender info
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) throw new Error("User not found");

  const project = await prisma.project.findUnique({
    where: { id: notice.projectId },
  });

  const org = await prisma.organization.findFirst();

  // Determine GC contact email — prefer project-level, fall back to notice recipient
  const toEmail = project?.gcContactEmail || notice.recipientEmail;
  const toName = project?.gcContactName || notice.recipientName;

  const sentMethods: string[] = [];
  const failedMethods: string[] = [];
  const confirmation: DeliveryConfirmation = {};

  for (const method of methods) {
    try {
      if (method === "EMAIL" && toEmail) {
        const fromEmail = org?.replyToDomain || "noreply@efilo.ai";
        const fromName = `${user.name} via ${org?.name || "efilo.ai"}`;

        await sendEmail({
          to: toEmail,
          toName: toName ?? undefined,
          from: fromEmail,
          fromName,
          replyTo: user.email,
          cc: user.email,
          subject: `NOTICE: ${notice.title}`,
          html: `<pre style="font-family: 'DM Sans', sans-serif; white-space: pre-wrap; max-width: 700px;">${escapeHtml(notice.content)}</pre>`,
          text: notice.content,
        });
        confirmation.email = {
          sentAt: new Date().toISOString(),
          status: "sent",
        };
        sentMethods.push(method);
      } else if (method === "CERTIFIED_MAIL") {
        // Manual tracking — user enters tracking number later
        confirmation.certifiedMail = {
          sentAt: new Date().toISOString(),
          trackingNumber: "PENDING_USER_INPUT",
          carrier: "USPS",
          status: "pending",
        };
        sentMethods.push(method);
      } else if (method === "HAND_DELIVERY") {
        confirmation.handDelivery = {
          deliveredAt: "PENDING_CONFIRMATION",
          receivedBy: "PENDING_CONFIRMATION",
        };
        sentMethods.push(method);
      } else if (method === "COURIER") {
        confirmation.courier = {
          sentAt: new Date().toISOString(),
          trackingNumber: "PENDING_USER_INPUT",
          carrier: "PENDING_USER_INPUT",
          status: "pending",
        };
        sentMethods.push(method);
      } else {
        // REGISTERED_MAIL, FAX, etc. — track as pending
        sentMethods.push(method);
      }
    } catch (error) {
      console.error(`Failed to send via ${method}:`, error);
      failedMethods.push(method);
    }
  }

  const newStatus: ComplianceNoticeStatus =
    sentMethods.length > 0 ? "SENT" : "DRAFT";

  await prisma.complianceNotice.update({
    where: { id: noticeId },
    data: {
      status: newStatus,
      sentAt: sentMethods.length > 0 ? new Date() : null,
      deliveryConfirmation: JSON.parse(JSON.stringify(confirmation)),
      deliveryMethods: sentMethods,
    },
  });

  // Update linked deadline status
  if (sentMethods.length > 0) {
    const deadline = await prisma.complianceDeadline.findFirst({
      where: { noticeId },
    });
    if (deadline) {
      await prisma.complianceDeadline.update({
        where: { id: deadline.id },
        data: { status: "NOTICE_SENT" },
      });
    }
  }

  await prisma.complianceAuditLog.create({
    data: {
      projectId: notice.projectId,
      eventType: "NOTICE_SENT",
      entityType: "ComplianceNotice",
      entityId: noticeId,
      userId,
      actorType: "USER",
      action: "sent",
      details: JSON.parse(JSON.stringify({ sentMethods, failedMethods })),
    },
  });

  return { sentMethods, failedMethods };
}

// ── Confirm Delivery ──────────────────────────────────────────────────────

interface ConfirmDeliveryParams {
  method: string;
  trackingNumber?: string;
  carrier?: string;
  deliveredAt?: string;
  signedBy?: string;
  receivedBy?: string;
}

export async function confirmDelivery(
  noticeId: string,
  params: ConfirmDeliveryParams,
  userId: string
): Promise<ComplianceNotice> {
  const notice = await prisma.complianceNotice.findUnique({
    where: { id: noticeId },
  });
  if (!notice) throw new Error("Notice not found");

  const existing =
    (notice.deliveryConfirmation as DeliveryConfirmation | null) || {};
  const now = new Date().toISOString();

  switch (params.method) {
    case "EMAIL":
      existing.email = {
        ...existing.email!,
        status: "delivered",
        deliveredAt: params.deliveredAt || now,
      };
      break;
    case "CERTIFIED_MAIL":
      existing.certifiedMail = {
        sentAt: existing.certifiedMail?.sentAt || now,
        trackingNumber: params.trackingNumber || "",
        carrier: params.carrier || "USPS",
        status: "delivered",
        deliveredAt: params.deliveredAt || now,
        signedBy: params.signedBy,
      };
      break;
    case "HAND_DELIVERY":
      existing.handDelivery = {
        deliveredAt: params.deliveredAt || now,
        receivedBy: params.receivedBy || "",
      };
      break;
    case "COURIER":
      existing.courier = {
        sentAt: existing.courier?.sentAt || now,
        trackingNumber: params.trackingNumber || "",
        carrier: params.carrier || "",
        status: "delivered",
        deliveredAt: params.deliveredAt || now,
      };
      break;
  }

  // Check if all methods are fully confirmed
  const allDelivered = notice.deliveryMethods.every((m) => {
    switch (m) {
      case "EMAIL":
        return existing.email?.status === "delivered";
      case "CERTIFIED_MAIL":
        return existing.certifiedMail?.status === "delivered";
      case "HAND_DELIVERY":
        return (
          existing.handDelivery?.receivedBy !== undefined &&
          existing.handDelivery.receivedBy !== "PENDING_CONFIRMATION"
        );
      case "COURIER":
        return existing.courier?.status === "delivered";
      default:
        return true;
    }
  });

  const onTimeStatus =
    allDelivered && notice.sentAt && notice.dueDate
      ? notice.sentAt <= notice.dueDate
      : null;

  const newStatus: ComplianceNoticeStatus = allDelivered
    ? "ACKNOWLEDGED"
    : "SENT";

  const updated = await prisma.complianceNotice.update({
    where: { id: noticeId },
    data: {
      deliveryConfirmation: JSON.parse(JSON.stringify(existing)),
      status: newStatus,
      deliveredAt: allDelivered ? new Date() : null,
      onTimeStatus,
    },
  });

  // Mark deadline as completed if all delivered
  if (allDelivered) {
    const deadline = await prisma.complianceDeadline.findFirst({
      where: { noticeId },
    });
    if (deadline) {
      await prisma.complianceDeadline.update({
        where: { id: deadline.id },
        data: { status: "COMPLETED" },
      });
    }
  }

  await prisma.complianceAuditLog.create({
    data: {
      projectId: notice.projectId,
      eventType: "DELIVERY_CONFIRMED",
      entityType: "ComplianceNotice",
      entityId: noticeId,
      userId,
      actorType: "USER",
      action: "delivery_confirmed",
      details: JSON.parse(
        JSON.stringify({
          method: params.method,
          allDelivered,
          onTimeStatus,
        })
      ),
    },
  });

  return updated;
}

// ── Query ─────────────────────────────────────────────────────────────────

export async function getProjectNotices(
  projectId: string,
  filters?: {
    type?: ComplianceNoticeType;
    status?: ComplianceNoticeStatus;
  }
) {
  return prisma.complianceNotice.findMany({
    where: {
      projectId,
      ...(filters?.type && { type: filters.type }),
      ...(filters?.status && { status: filters.status }),
    },
    orderBy: { createdAt: "desc" },
  });
}

function escapeHtml(str: string): string {
  return str
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

File: backend/app/services/compliance/notices.py
"""Notice management service.

Handles the full lifecycle of compliance notices: create, edit, send,
regenerate, confirm delivery, and delete.
"""

import logging
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.compliance import (
    ComplianceAuditLog,
    ComplianceDeadline,
    ComplianceNotice,
    ContractClause,
)
from app.models.enums import ComplianceNoticeStatus, ComplianceNoticeType, DeadlineStatus
from app.models.project import Project
from app.models.user import User
from app.services.ai import generate_response
from app.services.email import send_rfi_email

from .prompts import NOTICE_GENERATION_SYSTEM, NOTICE_GENERATION_USER

logger = logging.getLogger(__name__)


async def create_notice(
    db: AsyncSession,
    project_id: str,
    notice_type: ComplianceNoticeType,
    title: str,
    content: str,
    user_id: str,
    clause_id: str | None = None,
    due_date: datetime | None = None,
    recipient_name: str | None = None,
    recipient_email: str | None = None,
    deadline_id: str | None = None,
) -> ComplianceNotice:
    """Create a new compliance notice (draft)."""
    notice = ComplianceNotice(
        project_id=project_id,
        type=notice_type,
        status=ComplianceNoticeStatus.DRAFT,
        title=title,
        content=content,
        clause_id=clause_id,
        due_date=due_date,
        recipient_name=recipient_name,
        recipient_email=recipient_email,
        created_by_id=user_id,
    )
    db.add(notice)
    await db.flush()
    await db.refresh(notice)

    # If linked to a deadline, update deadline status
    if deadline_id:
        result = await db.execute(
            select(ComplianceDeadline).where(
                ComplianceDeadline.id == deadline_id,
                ComplianceDeadline.project_id == project_id,
            )
        )
        deadline = result.scalar_one_or_none()
        if deadline:
            deadline.status = DeadlineStatus.NOTICE_DRAFTED
            deadline.notice_id = notice.id
            deadline.notice_created_at = datetime.utcnow()

    # Audit log
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="NOTICE_CREATED",
        entity_type="ComplianceNotice",
        entity_id=notice.id,
        user_id=user_id,
        actor_type="USER",
        action="create_notice",
        details={
            "type": notice_type.value,
            "title": title,
            "deadlineId": deadline_id,
        },
    )
    db.add(audit)
    await db.flush()

    return notice


async def generate_notice_draft(
    db: AsyncSession,
    project_id: str,
    clause_id: str,
    trigger_description: str,
    trigger_date: datetime,
    deadline_date: datetime,
    notice_type: ComplianceNoticeType,
    user_id: str,
    additional_context: str = "",
) -> dict:
    """Generate a notice draft using Claude AI.

    Returns dict with content, model, tokensUsed.
    """
    # Load clause
    result = await db.execute(
        select(ContractClause).where(ContractClause.id == clause_id)
    )
    clause = result.scalar_one_or_none()
    if not clause:
        raise ValueError(f"Clause {clause_id} not found")

    # Load project
    result = await db.execute(
        select(Project).where(Project.id == project_id)
    )
    project = result.scalar_one_or_none()
    if not project:
        raise ValueError(f"Project {project_id} not found")

    # Load user
    result = await db.execute(
        select(User).where(User.id == user_id)
    )
    user = result.scalar_one_or_none()

    # Get GC contact from project
    gc_name = project.gc_contact_name or "General Contractor"
    gc_company = project.gc_company_name or ""
    gc_email = project.gc_contact_email or ""

    # Format notice type for display
    notice_type_display = notice_type.value.replace("_", " ").title()

    user_prompt = NOTICE_GENERATION_USER.format(
        notice_type=notice_type_display,
        project_name=project.name,
        clause_title=clause.title,
        clause_section_ref=clause.section_ref or "N/A",
        clause_content=clause.content,
        trigger_description=trigger_description,
        trigger_date=trigger_date.strftime("%B %d, %Y"),
        deadline_date=deadline_date.strftime("%B %d, %Y"),
        notice_method=clause.notice_method.value if clause.notice_method else "WRITTEN_NOTICE",
        from_name=user.name if user else "Project Manager",
        from_company=project.name,  # Organization name not directly on project
        to_name=gc_name,
        to_company=gc_company,
        to_email=gc_email,
        additional_context=additional_context or "None",
    )

    ai_response = generate_response(
        system_prompt=NOTICE_GENERATION_SYSTEM,
        user_prompt=user_prompt,
        model="sonnet",
        max_tokens=4000,
        temperature=0.2,
    )

    return {
        "content": ai_response.content,
        "model": ai_response.model,
        "tokensUsed": ai_response.tokens_used,
    }


async def update_notice(
    db: AsyncSession,
    notice_id: str,
    project_id: str,
    user_id: str,
    title: str | None = None,
    content: str | None = None,
    recipient_name: str | None = None,
    recipient_email: str | None = None,
    due_date: datetime | None = None,
) -> ComplianceNotice | None:
    """Update a notice (only DRAFT and PENDING_REVIEW can be edited)."""
    result = await db.execute(
        select(ComplianceNotice).where(
            ComplianceNotice.id == notice_id,
            ComplianceNotice.project_id == project_id,
        )
    )
    notice = result.scalar_one_or_none()
    if not notice:
        return None

    if notice.status not in (ComplianceNoticeStatus.DRAFT, ComplianceNoticeStatus.PENDING_REVIEW):
        raise ValueError(f"Cannot edit notice in {notice.status.value} status")

    if title is not None:
        notice.title = title
    if content is not None:
        notice.content = content
    if recipient_name is not None:
        notice.recipient_name = recipient_name
    if recipient_email is not None:
        notice.recipient_email = recipient_email
    if due_date is not None:
        notice.due_date = due_date

    await db.flush()
    return notice


async def send_notice(
    db: AsyncSession,
    notice_id: str,
    project_id: str,
    user_id: str,
) -> ComplianceNotice | None:
    """Send a compliance notice via email.

    Updates status to SENT and records delivery info.
    """
    result = await db.execute(
        select(ComplianceNotice).where(
            ComplianceNotice.id == notice_id,
            ComplianceNotice.project_id == project_id,
        )
    )
    notice = result.scalar_one_or_none()
    if not notice:
        return None

    if notice.status not in (ComplianceNoticeStatus.DRAFT, ComplianceNoticeStatus.PENDING_REVIEW):
        raise ValueError(f"Cannot send notice in {notice.status.value} status")

    if not notice.recipient_email:
        raise ValueError("Notice has no recipient email")

    # Load project and user for email context
    project_result = await db.execute(
        select(Project).where(Project.id == project_id)
    )
    project = project_result.scalar_one_or_none()

    user_result = await db.execute(
        select(User).where(User.id == user_id)
    )
    user = user_result.scalar_one_or_none()

    # Send email
    from_name = user.name if user else "efilo.ai"
    from_email = f"noreply@efilo.ai"
    reply_to = user.email if user else from_email
    project_name = project.name if project else "Project"

    sent = send_rfi_email(
        from_name=from_name,
        from_email=from_email,
        reply_to=reply_to,
        to=notice.recipient_email,
        to_name=notice.recipient_name,
        cc=user.email if user else None,
        rfi_number=f"NOTICE-{notice.id[:8]}",
        subject=notice.title,
        question=notice.content,
        project_name=project_name,
    )

    now = datetime.utcnow()
    notice.status = ComplianceNoticeStatus.SENT
    notice.sent_at = now
    notice.delivered_at = now if sent else None
    notice.delivery_methods = ["EMAIL"]
    notice.on_time_status = notice.due_date is None or now <= notice.due_date

    # Update linked deadline
    if notice.clause_id:
        deadline_result = await db.execute(
            select(ComplianceDeadline).where(
                ComplianceDeadline.notice_id == notice.id,
                ComplianceDeadline.project_id == project_id,
            )
        )
        deadline = deadline_result.scalar_one_or_none()
        if deadline:
            deadline.status = DeadlineStatus.NOTICE_SENT

    # Audit log
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="NOTICE_SENT",
        entity_type="ComplianceNotice",
        entity_id=notice_id,
        user_id=user_id,
        actor_type="USER",
        action="send_notice",
        details={
            "recipientEmail": notice.recipient_email,
            "emailSent": sent,
            "onTime": notice.on_time_status,
        },
    )
    db.add(audit)
    await db.flush()

    return notice


async def delete_notice(
    db: AsyncSession,
    notice_id: str,
    project_id: str,
    user_id: str,
) -> bool:
    """Delete a notice (only DRAFT and PENDING_REVIEW)."""
    result = await db.execute(
        select(ComplianceNotice).where(
            ComplianceNotice.id == notice_id,
            ComplianceNotice.project_id == project_id,
        )
    )
    notice = result.scalar_one_or_none()
    if not notice:
        return False

    if notice.status not in (ComplianceNoticeStatus.DRAFT, ComplianceNoticeStatus.PENDING_REVIEW):
        raise ValueError(f"Cannot delete notice in {notice.status.value} status")

    # Unlink from deadline if linked
    deadline_result = await db.execute(
        select(ComplianceDeadline).where(
            ComplianceDeadline.notice_id == notice_id,
            ComplianceDeadline.project_id == project_id,
        )
    )
    deadline = deadline_result.scalar_one_or_none()
    if deadline:
        deadline.notice_id = None
        deadline.notice_created_at = None
        deadline.status = DeadlineStatus.ACTIVE

    # Audit before delete
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="NOTICE_DELETED",
        entity_type="ComplianceNotice",
        entity_id=notice_id,
        user_id=user_id,
        actor_type="USER",
        action="delete_notice",
        details={"title": notice.title, "type": notice.type.value},
    )
    db.add(audit)

    await db.delete(notice)
    await db.flush()
    return True


async def get_notices(
    db: AsyncSession,
    project_id: str,
    status: ComplianceNoticeStatus | None = None,
    notice_type: ComplianceNoticeType | None = None,
) -> list[ComplianceNotice]:
    """List notices for a project."""
    query = (
        select(ComplianceNotice)
        .where(ComplianceNotice.project_id == project_id)
        .order_by(ComplianceNotice.created_at.desc())
    )

    if status:
        query = query.where(ComplianceNotice.status == status)
    if notice_type:
        query = query.where(ComplianceNotice.type == notice_type)

    result = await db.execute(query)
    return list(result.scalars().all())


async def confirm_delivery(
    db: AsyncSession,
    notice_id: str,
    project_id: str,
    user_id: str,
    method: str,
    tracking_number: str | None = None,
    carrier: str | None = None,
    delivered_at: datetime | None = None,
    signed_by: str | None = None,
    received_by: str | None = None,
) -> ComplianceNotice | None:
    """Confirm delivery of a sent notice with tracking info."""
    result = await db.execute(
        select(ComplianceNotice).where(
            ComplianceNotice.id == notice_id,
            ComplianceNotice.project_id == project_id,
        )
    )
    notice = result.scalar_one_or_none()
    if not notice:
        return None

    if notice.status != ComplianceNoticeStatus.SENT:
        raise ValueError(f"Cannot confirm delivery for notice in {notice.status.value} status")

    now = datetime.utcnow()
    existing = notice.delivery_confirmation or {}

    method_lower = method.lower().replace("_", "")
    confirmation_entry = {
        "status": "delivered",
        "deliveredAt": (delivered_at or now).isoformat(),
    }
    if tracking_number:
        confirmation_entry["trackingNumber"] = tracking_number
    if carrier:
        confirmation_entry["carrier"] = carrier
    if signed_by:
        confirmation_entry["signedBy"] = signed_by
    if received_by:
        confirmation_entry["receivedBy"] = received_by

    method_key_map = {
        "email": "email",
        "certifiedmail": "certifiedMail",
        "registeredmail": "registeredMail",
        "handdelivery": "handDelivery",
        "fax": "fax",
        "courier": "courier",
    }
    key = method_key_map.get(method_lower, method_lower)
    existing[key] = confirmation_entry

    notice.delivery_confirmation = existing
    notice.delivered_at = delivered_at or now
    notice.status = ComplianceNoticeStatus.ACKNOWLEDGED

    # Add delivery method if not already present
    methods = list(notice.delivery_methods or [])
    if method not in methods:
        methods.append(method)
        notice.delivery_methods = methods

    # Audit
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="DELIVERY_CONFIRMED",
        entity_type="ComplianceNotice",
        entity_id=notice_id,
        user_id=user_id,
        actor_type="USER",
        action="confirm_delivery",
        details={"method": method, "trackingNumber": tracking_number},
    )
    db.add(audit)
    await db.flush()

    return notice


async def regenerate_notice_draft(
    db: AsyncSession,
    notice_id: str,
    project_id: str,
    user_id: str,
    custom_instructions: str | None = None,
) -> dict:
    """Regenerate a notice's content using AI.

    Only DRAFT and PENDING_REVIEW notices can be regenerated.
    Returns the new content.
    """
    result = await db.execute(
        select(ComplianceNotice).where(
            ComplianceNotice.id == notice_id,
            ComplianceNotice.project_id == project_id,
        )
    )
    notice = result.scalar_one_or_none()
    if not notice:
        raise ValueError("Notice not found")

    if notice.status not in (ComplianceNoticeStatus.DRAFT, ComplianceNoticeStatus.PENDING_REVIEW):
        raise ValueError(f"Cannot regenerate notice in {notice.status.value} status")

    if not notice.clause_id:
        raise ValueError("Notice has no linked clause")

    # Load clause
    clause_result = await db.execute(
        select(ContractClause).where(ContractClause.id == notice.clause_id)
    )
    clause = clause_result.scalar_one_or_none()
    if not clause:
        raise ValueError("Linked clause not found")

    # Find linked deadline
    deadline_result = await db.execute(
        select(ComplianceDeadline).where(ComplianceDeadline.notice_id == notice_id)
    )
    deadline = deadline_result.scalar_one_or_none()

    # Load project
    project_result = await db.execute(
        select(Project).where(Project.id == project_id)
    )
    project = project_result.scalar_one_or_none()

    # Load user
    user_result = await db.execute(
        select(User).where(User.id == user_id)
    )
    user = user_result.scalar_one_or_none()

    gc_name = project.gc_contact_name or "General Contractor" if project else "General Contractor"
    gc_company = project.gc_company_name or "" if project else ""
    gc_email = project.gc_contact_email or "" if project else ""

    trigger_date = deadline.triggered_at if deadline else datetime.utcnow()
    deadline_date = deadline.calculated_deadline if deadline else notice.due_date or datetime.utcnow()

    additional = custom_instructions or "None"
    notice_type_display = notice.type.value.replace("_", " ").title()

    user_prompt = NOTICE_GENERATION_USER.format(
        notice_type=notice_type_display,
        project_name=project.name if project else "Project",
        clause_title=clause.title,
        clause_section_ref=clause.section_ref or "N/A",
        clause_content=clause.content,
        trigger_description=deadline.trigger_description if deadline else "Manual notice",
        trigger_date=trigger_date.strftime("%B %d, %Y"),
        deadline_date=deadline_date.strftime("%B %d, %Y"),
        notice_method=clause.notice_method.value if clause.notice_method else "WRITTEN_NOTICE",
        from_name=user.name if user else "Project Manager",
        from_company=project.name if project else "",
        to_name=gc_name,
        to_company=gc_company,
        to_email=gc_email,
        additional_context=additional,
    )

    ai_response = generate_response(
        system_prompt=NOTICE_GENERATION_SYSTEM,
        user_prompt=user_prompt,
        model="sonnet",
        max_tokens=4000,
        temperature=0.2,
    )

    # Update notice with new content
    notice.content = ai_response.content
    notice.generated_by_ai = True
    notice.ai_model = ai_response.model
    await db.flush()

    return {"content": ai_response.content}


async def get_notice_by_id(
    db: AsyncSession,
    notice_id: str,
    project_id: str,
) -> ComplianceNotice | None:
    """Get a single notice."""
    result = await db.execute(
        select(ComplianceNotice).where(
            ComplianceNotice.id == notice_id,
            ComplianceNotice.project_id == project_id,
        )
    )
    return result.scalar_one_or_none()
